### ISSUE TITLE
Changing note attributes in collab setting and undoing leads to node duplication

### ISSUE THREAD

**Issue by matej-svejda** [ProseMirror/prosemirror#939]
### Issue details
Im not sure if this is intended behaviour or not, but it seems strange to me:

When collaborating on a document and moving a note on client A, then changing an attribute on the same node (i.e. replacing the node) on client B and then undoing the step in client A, the node is duplicated: it appears at the original position with the original attribute and at the moved to position with the changed attribute.

The duplication doesn't happen when you move the node into a node where it doesn't belong (paragraph for example) using tr.insert, which then triggers replaceStep that creates a two-way open Slice for it. I wasn't able to trigger this operation in the collab example (you can't seem to drop nodes into nodes where they don't belong there).

### Steps to reproduce
In the collab example on the website you can reproduce it for headings for example.

### ProseMirror version
prosemirror-history 1.0.4 and prosemirror-collab 1.1.1

### Affected platforms
Pretty sure its browser independent, I tested in on Chrome/Firefox on Mac/Linux.

**Comment by marijnh** [comment]
I think the right thing to do would be to provide a separate step type for attribute changes, and have it (like mark steps) return an empty `StepMap` so that changing a node's attributes isn't treated like replacing the node.

You can do this in user code for the time being. I'll try to write up an RFC to add it to prosemirror-transform in the near future.

**Comment by matej-svejda** [comment]
Good point, I didnt think in the direction of creating a new step type for taht. Ill maybe end up creating a specific step for my case, since Im "missusing" attributes a bit to pass data to a custom NodeView and could use the more fine-grained control I get :)

**Comment by marijnh** [comment]
I ran into some issues with my design (specifically, if you have code tracking changes by looking at step maps, it'd miss changes created by such steps, possibly breaking existing user code), so I'm not sure what the RFC should look like yet. Since I have a lot of stuff to do this week, I'm going to put off working on this for a bit.

**Comment by matej-svejda** [comment]
I implemented my own step now which can take advantage of some assumptions specific to my usecase, so no pressure from me :)
It didnt occur to me that there might be code relying on StepMap being non-empty. Is it only user code or are you aware of parts of PM that might have this assumption?

**Comment by marijnh** [comment]
PM itself should be fine, and probably most user code as well. But for example if you have an id-assigning addon that makes sure every node of a given type has a unique id attr, and it works by checking for changed ranges and checking only those, this change would break that.

Still debating whether that is unlikely enough to be acceptable or not.

**Comment by marijnh** [comment]
This issue has been moved to [issue #939](https://code.haverbeke.berlin/prosemirror/prosemirror/issues/939) on the [repository's new home](https://code.haverbeke.berlin/prosemirror/prosemirror/).
